### ISSUE TITLE
Incorrect get_num_losses return value when using custom loss function

### ISSUE THREAD

**Issue by zinccat** [NVIDIA/modulus-sym#24]
The implementation of get_num_losses in `domain.py` is summing up the number of output names in constraints. Yet this is not the case when using custom loss function, e.g. when using variational constraint like the case in https://docs.nvidia.com/deeplearning/modulus/modulus-sym/user_guide/intermediate/variational_example.html

**Comment by NickGeneva** [comment]
Thanks for the report, I agree this is a bug. The `get_num_losses` in [domain](https://github.com/NVIDIA/modulus-sym/blob/main/modulus/sym/domain/domain.py#L128) uses the output names of each constraint to get a total number of losses.

The custom loss, particularly in the variational constraint does not give its output variables ahead of time and the variational constraint [constructor](https://github.com/NVIDIA/modulus-sym/blob/b06727277c67fee8a9e291ce3e7b19db5505eb85/modulus/sym/domain/constraint/continuous.py#L833) will have 0 output keys (and thus say it has zero losses) since there are no output keys of a variational data set.

Unsure about a clean fix, without requiring additional user information its not possible presently to tell that a custom loss exists in this constraint without a forward pass. Example [impacted](https://github.com/NVIDIA/modulus-sym/blob/main/examples/discontinuous_galerkin/dg/dg.py#L146C33).

**Comment by CANCEL-FEM** [comment]
I have run into this issue with both ResNorm and GradNorm aggregators when using custom loss functions. It looks like they are set up mainly for PDEs, where the number of outputs matches the number of losses. Anyway, here is a quick workaround if you are working with custom loss functions: you need just to hardcode the aggregator class, like I did with ResNorm below (same goes for GradNorm). You can also tweak the class to inherit num_losses from the config file if that works better for you. Hope this helps.

class FixedResNorm(Aggregator):

    def __init__(self, params, num_losses, alpha=1.0, weights=None):
        #######Here you force the number of losses and ignore the one passed to this class#######
        forced_num_losses = 2  
        
        super().__init__(params, forced_num_losses, weights)
        self.alpha: float = alpha
        

        self.lmbda: torch.nn.Parameter = nn.Parameter(
            torch.zeros(forced_num_losses, device=self.device)
        )
        

        self.register_buffer(
            "init_losses", torch.zeros(forced_num_losses, device=self.device)
        )

    def forward(self, losses: Dict[str, torch.Tensor], step: int) -> torch.Tensor:
        # weigh losses
        losses = self.weigh_losses(losses, self.weights)

        # get initial losses
        if step == 0:
            #######Safety Check#######
            if len(losses) != self.num_losses:
                print(f"Error: FixedResNorm expects {forced_num_losses}  losses, but got {len(losses)}: {list(losses.keys())}")

            for i, key in enumerate(losses.keys()):
                self.init_losses[i] = losses[key].clone().detach()

        with torch.no_grad():
            normalizer: torch.Tensor = self.num_losses / (torch.exp(self.lmbda).sum())
            for i in range(self.num_losses):
                self.lmbda[i] = self.lmbda[i].clone() + torch.log(
                    normalizer.detach()
                )  # c*exp(x) = exp(log(c)+x)
        lmbda_exp: torch.Tensor = torch.exp(self.lmbda)

        # compute relative losses
        losses_stacked: torch.Tensor = torch.stack(list(losses.values()))
        
        with torch.no_grad():

            relative_losses: torch.Tensor = torch.div(losses_stacked, self.init_losses)
            inverse_rate: torch.Tensor = relative_losses / (relative_losses.mean())
            resnorm_coef: torch.Tensor = torch.pow(inverse_rate, self.alpha)

        # compute residual norm
        residuals: torch.Tensor = torch.zeros_like(self.init_losses)
        for i, key in enumerate(losses.keys()):
            residuals[i] = lmbda_exp[i] * losses[key].detach()
        avg_residuals: torch.Tensor = losses_stacked.detach().mean()

        # compute final loss
        loss_resnorm: torch.Tensor = torch.abs(
            residuals - avg_residuals * resnorm_coef
        ).sum()
        loss_model: torch.Tensor = (lmbda_exp.detach() * losses_stacked).sum()
        loss: torch.Tensor = loss_resnorm + loss_model
        
        return loss

**Comment by ktangsali** [comment]
Closing as part of the archival of this repository. The key components have been upstreamed to [NVIDIA/physicsnemo](https://github.com/NVIDIA/physicsnemo). Please refer to the [v2.0 Migration Guide](https://github.com/NVIDIA/physicsnemo/blob/main/v2.0-MIGRATION-GUIDE.md#physicsnemo-sym--physicsnemosym) and feel free to re-open your issue there if it is still relevant.
